### PR TITLE
(not so WIP anymore) Storagewhiff

### PR DESCRIPTION
**Pull request by mhhf**

### CODE REVIEW

File: src/hevm/src/EVM.hs [hunk 1]
```diff
@@ -178,7 +178,7 @@ data Cache = Cache
 -- | A way to specify an initial VM state
 data VMOpts = VMOpts
   { vmoptContract :: Contract
-  , vmoptCalldata :: (Buffer, (SWord 32)) -- maximum size of uint32 as per eip 1985
+  , vmoptCalldata :: (Buffer, SWord 256) -- maximum size of uint32 as per eip 1985
```
Comment (MrChico) [inline src/hevm/src/EVM.hs:181]:
```suggestion
  , vmoptCalldata :: (Buffer, SWord 256)
```

File: src/hevm/src/EVM/Types.hs [hunk 2]
```diff
@@ -45,35 +50,69 @@ import qualified Text.Read
...
+data Word = C Whiff W256 --maybe to remove completely in the future
+
+instance Show Word where
+  show (C _ x) = show x
+
+-- | Symbolic words of 256 bits, possibly annotated with additional
+--   "insightful" information
+data SymWord = S Whiff (SWord 256)
+
+instance Show SymWord where
```
Comment (MrChico) [inline src/hevm/src/EVM/Types.hs:73]: can you move the other instances of `SymWord` to here as well? Right now we are getting a lot of warnings about orphan instances

File: src/hevm/src/EVM/Types.hs [hunk 3]
```diff
@@ -83,6 +122,31 @@ class FromSizzleBV a where
...
+
+-- | Conversion from a fixed-sized BV to a sized bit-vector.
+class ToSizzleBV a where
+   -- | Convert a fixed-sized bit-vector to the corresponding sized bit-vector,
+   toSizzle :: a -> ToSizzle a
+
+   default toSizzle :: (Num (ToSizzle a), Integral a) => (a -> ToSizzle a)
+   toSizzle = fromIntegral
+
+
```
Comment (MrChico) [inline src/hevm/src/EVM/Types.hs:148]:
```suggestion
```

File: src/hevm/src/EVM/Types.hs [hunk 4]
```diff
@@ -83,6 +122,31 @@ class FromSizzleBV a where
...
 
+maybeLitWord :: SymWord -> Maybe Word
+maybeLitWord (S whiff a) = fmap (C whiff . fromSizzle) (unliteral a)
+
+
+-- We need a 512-bit word for doing ADDMOD and MULMOD with full precision.
+mkUnpackedDoubleWord "Word512" ''Word256 "Int512" ''Int256 ''Word256
+  [''Typeable, ''Data, ''Generic]
+
+
```
Comment (MrChico) [inline src/hevm/src/EVM/Types.hs:133]:
```suggestion
```

File: src/hevm/src/EVM/Types.hs [hunk 5]
```diff
@@ -83,6 +122,31 @@ class FromSizzleBV a where
    default fromSizzle :: (Num (FromSizzle a), Integral a) => a -> FromSizzle a
    fromSizzle = fromIntegral
 
+maybeLitWord :: SymWord -> Maybe Word
+maybeLitWord (S whiff a) = fmap (C whiff . fromSizzle) (unliteral a)
+
+
```
Comment (MrChico) [inline src/hevm/src/EVM/Types.hs:128]:
```suggestion
```
plz

File: src/hevm/src/EVM/TTY.hs [hunk 6]
```diff
@@ -12,13 +12,12 @@ import Brick.Widgets.List
...
 import EVM
 import EVM.ABI (abiTypeSolidity, decodeAbiValue, AbiType(..), emptyAbi)
-import EVM.Symbolic (SymWord(..))
 import EVM.SymExec (maxIterationsReached, symCalldata)
 import EVM.Dapp (DappInfo, dappInfo, Test, extractSig, Test(..))
 import EVM.Dapp (dappUnitTests, unitTestMethods, dappSolcByName, dappSolcByHash, dappSources)
 import EVM.Dapp (dappAstSrcMap)
 import EVM.Debug
-import EVM.Format (showWordExact, showWordExplanation)
+import EVM.Format (showWordExact, showWordExplanation )
```
Comment (MrChico) [inline src/hevm/src/EVM/TTY.hs:20]:
```suggestion
import EVM.Format (showWordExact, showWordExplanation)
```

File: src/hevm/src/EVM/UnitTest.hs [hunk 7]
```diff
@@ -763,8 +763,11 @@ symbolify vm =
      & set (env . storageModel) InitialS
   where
     mkSymStorage :: Storage -> Storage
-    mkSymStorage (Symbolic _) = error "should not happen"
-    mkSymStorage (Concrete s) = Symbolic $ sListArray 0 [(literal $ toSizzle k, v) | (C _ k, S _ v) <- Map.toList s]
+    mkSymStorage (Symbolic _ _) = error "should not happen"
+    mkSymStorage (Concrete s) = let
+        list = [(literal $ toSizzle k, v) | (C _ k, S _ v) <- Map.toList s]
```
Comment (MrChico) [inline src/hevm/src/EVM/UnitTest.hs:768]:
```suggestion
    mkSymStorage (Concrete s) = 
      let
        list = [(literal $ toSizzle k, v) | (C _ k, S _ v) <- Map.toList s]
```

**Comment by MrChico** [comment]
Changelog entry and this looks good

File: src/hevm/src/EVM/Concrete.hs [hunk 8]
```diff
@@ -115,20 +90,9 @@ setMemoryByte :: Word -> Word8 -> ByteString -> ByteString
...
-blobSize :: ByteString -> Word
-blobSize x = w256 (num (BS.length x))
-
 keccakBlob :: ByteString -> Word
 keccakBlob x = C (FromKeccak x) (keccak x)
 
-instance Show Word where
-  show (C _ x) = show x
 
 instance Read Word where
```
Comment (MrChico) [inline src/hevm/src/EVM/Concrete.hs:97]: also these should go in `Types.hs` now

File: src/hevm/src/EVM/Format.hs [hunk 9]
```diff
@@ -336,48 +335,154 @@ prettyvmresult (EVM.VMSuccess (ConcreteBuffer msg)) =
...
+showBranchTree srcInfo (Node _ children) =
+  let
+    treeLines = flattenForest srcInfo (restack (flattenAbi False children))
+    doMax treeLine x = max x $ length $ _indent treeLine
+    maxIndent = 2 + foldr doMax 0 treeLines
+    showTreeLine bd = let
+      colIndent  = _indent bd
+      colContent = _content bd
+      indentSize = (maxIndent -(length colIndent))
+      in colIndent <> leftpad indentSize colContent
```
Comment (MrChico) [inline src/hevm/src/EVM/Format.hs:443]:
```suggestion
    showTreeLine (TreeLine colIndent colContent) =
      let
        indentSize = maxIndent - length colIndent
      in colIndent <> leftpad indentSize colContent
```

File: src/hevm/src/EVM/Format.hs [hunk 10]
```diff
@@ -336,48 +335,154 @@ prettyvmresult (EVM.VMSuccess (ConcreteBuffer msg)) =
...
+
+
+data AbiBranching = NoBranching | ConcreteBranching | CompareBranching
+
+abiBranching :: BranchInfo -> AbiBranching
+abiBranching (BranchInfo _ (Just (UnOp "isZero" (InfixBinOp "==" (Val _) (FromCalldata _ _))))) = ConcreteBranching
+abiBranching (BranchInfo _ (Just (UnOp "isZero" (InfixBinOp ">" (Val _) (FromCalldata _ _))))) = CompareBranching
+abiBranching _ = NoBranching
+
+flattenAbi :: Bool ->               -- have we hit the abi barrier?
```
Comment (MrChico) [inline src/hevm/src/EVM/Format.hs:454]:
```suggestion
flattenAbi :: Bool ->               -- ^ have we hit the abi barrier?
```

Comment (MrChico) [reply src/hevm/src/EVM/Format.hs:454]: this is haddoc format

File: src/hevm/src/EVM/Format.hs [hunk 11]
```diff
@@ -336,48 +335,154 @@ prettyvmresult (EVM.VMSuccess (ConcreteBuffer msg)) =
...
+              [Tree BranchInfo] ->
+              [Tree BranchInfo]
+flattenAbi False forest@((Node lbi ls):(Node rbi rs):_) = let
+    ff = flattenAbi False
+    ft = flattenAbi True
+  in case abiBranching lbi of
+      NoBranching       -> map (over branches ff) forest
```
Comment (MrChico) [inline src/hevm/src/EVM/Format.hs:461]:
```suggestion
      NoBranching       -> over branches ff <$> forest
```

File: src/hevm/src/EVM/Format.hs [hunk 12]
```diff
@@ -336,48 +335,154 @@ prettyvmresult (EVM.VMSuccess (ConcreteBuffer msg)) =
...
+abiForest [] = []
+abiForest f@((Node bi _):_) = let
+  xx = concatMap (abiForest . view branches) f
+  in if isAbi bi then f else xx
+
+-- TODO - doesn't work correctly, compare to real tree
+restack :: [Tree BranchInfo] -> [Tree BranchInfo]
+restack forest = let
+  insertIntoHollow cs [] = cs
+  insertIntoHollow cs f@((Node bi cl):xs) = if isAbi bi then cs else map (over branches (insertIntoHollow cs)) f
```
Comment (MrChico) [inline src/hevm/src/EVM/Format.hs:486]:
```suggestion
abiForest :: [Tree BranchInfo] -> [Tree BranchInfo]
abiForest [] = []
abiForest f@((Node bi _):_) =
  case abiBranching bi of
      ConcreteBranching -> f
      _ -> concatMap (abiForest . view branches) f

-- TODO - doesn't work correctly, compare to real tree
restack :: [Tree BranchInfo] -> [Tree BranchInfo]
restack forest = let
  insertIntoHollow cs [] = cs
  insertIntoHollow cs f@((Node bi cl):xs) = 
  case abiBranching bi of 
     ConcreteBranching -> cs 
     _ -> over branches (insertIntoHollow cs) <$> f
```

File: src/hevm/src/EVM/Format.hs [hunk 13]
```diff
@@ -336,48 +335,154 @@ prettyvmresult (EVM.VMSuccess (ConcreteBuffer msg)) =
...
+      NoBranching       -> map (over branches ff) forest
+      ConcreteBranching -> (Node rbi rs) : (ft ls)
+      CompareBranching  -> (ft ls) ++ (ft rs)
+flattenAbi True forest@((Node lbi ls):(Node rbi rs):_) = let
+    ft = flattenAbi True
+  in case abiBranching lbi of
+      NoBranching       -> forest
+      ConcreteBranching -> (Node rbi rs) : (ft ls)
+      CompareBranching  -> (ft ls) ++ (ft rs)
+flattenAbi _ t = t -- should never happen
```
Comment (MrChico) [inline src/hevm/src/EVM/Format.hs:470]: should we make this an error then?

File: src/hevm/src/EVM/Format.hs [hunk 14]
```diff
@@ -336,48 +335,154 @@ prettyvmresult (EVM.VMSuccess (ConcreteBuffer msg)) =
...
-data BranchData = BranchData {
-  _navigation :: String,
-  _constraint :: String,
-  _leafData :: String
+--
+-- TreeLine describes a singe line of the tree
+-- it contains the indentation which is prefixed to it
+-- and its content which contains the rest
+data TreeLine = TreeLine {
+  _indent   :: String,
```
Comment (MrChico) [inline src/hevm/src/EVM/Format.hs:353]: will this always just be a bunch of spaces? Maybe int is a better type for it then?

Comment (mhhf) [reply src/hevm/src/EVM/Format.hs:353]: no, its the line segment of the tree. e.g. it will be ` |    | |      | `

Comment (mhhf) [reply src/hevm/src/EVM/Format.hs:353]: immagine different symbols with different spacing
stupid github markdown screws it up

Comment (MrChico) [reply src/hevm/src/EVM/Format.hs:353]: I see. Do you think you have reimplemented [showTree](https://hackage.haskell.org/package/tree-view-0.5/docs/Data-Tree-View.html) here?

File: src/hevm/src/EVM/Format.hs [hunk 15]
```diff
@@ -353,94 +353,94 @@ data TreeLine = TreeLine {
...
-showTreeIndentSymbol False False = "\x2502"
+showTreeIndentSymbol False False = "\x2502" -- │
+
+flattenTree :: Int -> -- total number of cases
+               Int -> -- case index
+               Tree [String] ->
+               [TreeLine]
+
+-- this case should never happen for our use case, here for generality
+-- flattenTree _ _ (Node [] []) = []
```
Comment (MrChico) [inline src/hevm/src/EVM/Format.hs:376]:
```suggestion
```

**Pull request merged**